File: app/src/main/java/com/snhu/ProjectTwo/fragments/DatabaseGridFragment.java
package com.snhu.ProjectTwo.fragments;

import android.Manifest;
import android.app.DatePickerDialog;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.telephony.SmsManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.snhu.ProjectTwo.R;
import com.snhu.ProjectTwo.activities.CoreApp;
import com.snhu.ProjectTwo.interfaces.OnItemActionListener;
import com.snhu.ProjectTwo.utilities.DatabaseAdapter;
import com.snhu.ProjectTwo.utilities.LoginDatabase;
import com.snhu.ProjectTwo.utilities.UserInfo;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

//@Author Christian Clark
//@Date 8-14-25

//This fragment contains the database grid and is also where a user inputs their current weight of that moment
public class DatabaseGridFragment extends Fragment {

    private DatabaseAdapter _adapter;
    private List<UserInfo> _list;
    long _userId;

    private RecyclerView _recyclerView;
    private EditText _weightEntry;
    private ImageButton _addButton;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceData){
        return inflater.inflate(R.layout.database_grid_recycle_fragment, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState){
        super.onViewCreated(view, savedInstanceState);

        _recyclerView = view.findViewById(R.id.recyclerView);
        _weightEntry = view.findViewById(R.id.weight_entry_database);
        _addButton = view.findViewById(R.id.weight_add_button);

        _userId = ((CoreApp)getActivity()).getId();

        InitData();

        setupRecyclerView();
        setupAddButton();
    }

    private void InitData(){
        LoginDatabase db = new LoginDatabase(getContext());
        try
        {
            _list = db.GetInfoListByUser(_userId);
        }
        catch (Exception e)
        {
            Toast.makeText(getContext(), "Could not get user data", Toast.LENGTH_SHORT).show();
        }
    }

    //Add current weight button
    private void setupAddButton(){
        _addButton.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v){
                try{
                    float currWeight = Float.parseFloat(_weightEntry.getText().toString());
                    _weightEntry.setText("");
                    if(Float.compare(currWeight, 0.0f) == 0 || currWeight < 0){
                        Toast.makeText(getContext(), "Cannot have negative or 0 weight", Toast.LENGTH_SHORT).show();
                        return;
                    }
                    if(currWeight >= 3000f){
                        Toast.makeText(getContext(), "Weight too high", Toast.LENGTH_SHORT).show();
                        return;
                    }
                    SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
                    String dateString = dateFormat.format(new Date());

                    //adds the current weight to the database
                    LoginDatabase db = new LoginDatabase(getContext());
                    db.AddNewWeight(_userId, dateString, currWeight);
                    _list = db.GetInfoListByUser(_userId);
                    _adapter.updateData(_list);

                    //functionality for sending SMS messages to the user when they reach within 10 pounds or pass their goal
                    float goalWeight = db.GetGoalWeight(_userId);
                    if(_list.isEmpty()){
                        return;
                    }
                    currWeight = _list.get(0).getWeight();
                    String emulatorPort = "5554";
                    float weightToGo = 0;
                    AlertType alert = AlertType.Dont_send;
                    if(currWeight > goalWeight){
                        weightToGo = currWeight - goalWeight;
                        if(weightToGo <= 10.0f){
                            alert = AlertType.Close;
                        }
                    }else if(currWeight <= goalWeight){
                        alert = AlertType.Reached_Goal;
                    }
                    if(alert != AlertType.Dont_send){
                        TrySendSMS(alert, emulatorPort, weightToGo);
                    }
                }catch (Exception e){
                    Toast.makeText(getContext(), "Invalid Weight Entry", Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    //sets up the recycler view so users can scroll through their list of weights
    private void setupRecyclerView(){
        _recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        _adapter = new DatabaseAdapter(getContext(), _list, _userId, new OnItemActionListener() {
            //when clicking the delete button on a card
            @Override
            public void onDeleteClicked(UserInfo info, int position) {
                LoginDatabase db = new LoginDatabase(getContext());
                db.RemoveWeightAt(info.getId(), _userId);

                _adapter.updateData(db.GetInfoListByUser(_userId));
            }

            //when clicking the date button on a card
            @Override
            public void onDateClicked(UserInfo info, int position) {
                Calendar calendar = Calendar.getInstance();
                DatePickerDialog datePickerDialog = new DatePickerDialog(
                        getContext(),
                        (view, year, month, dayOfMonth) -> {
                            String date = year + "-" + (month + 1) + "-" + dayOfMonth;
                            LoginDatabase db = new LoginDatabase(getContext());
                            Toast.makeText(getContext(), "date: " + date, Toast.LENGTH_SHORT).show();
                            db.ChangeDateAt(info.getId(), _userId, date);
                            _adapter.updateData(db.GetInfoListByUser(_userId));
                        },
                        calendar.get(Calendar.YEAR),calendar.get(Calendar.MONTH),calendar.get(Calendar.DAY_OF_MONTH)
                );
                datePickerDialog.show();
            }
        });
        _recyclerView.setAdapter(_adapter);


        _recyclerView.setHasFixedSize(true);
    }

    //attempts to send an SMS message
    public void TrySendSMS(AlertType type, String phoneNumber, float poundsToGo){
        if(ContextCompat.checkSelfPermission(getContext(), Manifest.permission.SEND_SMS) == PackageManager.PERMISSION_GRANTED){
            SmsManager manager = getContext().getSystemService(SmsManager.class);
            String message;
            switch (type){
                case Close:
                    message = "You're only " + poundsToGo + " pounds from passing your goal!";
                    manager.sendTextMessage(phoneNumber, null, message, null, null);
                    break;
                case Reached_Goal:
                    message = "You've passed your weight goal!";
                    manager.sendTextMessage(phoneNumber, null, message, null, null);
                    break;
                default:
                    Toast.makeText(getContext(), "Something went horribly wrong", Toast.LENGTH_SHORT).show();
                    break;
            }
        }
    }

    //small enum for storing the type of sms alert to send
    private enum AlertType{
        Close,
        Reached_Goal,
        Dont_send
    }

}

File: app/src/main/java/com/snhu/ProjectTwo/fragments/DatabaseGridFragment.kt
package com.snhu.ProjectTwo.fragments

import android.Manifest
import android.app.DatePickerDialog
import android.app.NotificationManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.snhu.ProjectTwo.R
import com.snhu.ProjectTwo.activities.CoreApp
import com.snhu.ProjectTwo.databinding.DatabaseGridRecycleFragmentBinding
import com.snhu.ProjectTwo.entities.GoalEntity
import com.snhu.ProjectTwo.entities.UserInfoEntity
import com.snhu.ProjectTwo.utilities.AddNewWeight
import com.snhu.ProjectTwo.utilities.WeightCardAdapter
import com.snhu.ProjectTwo.utilities.WeightDatabase
import com.snhu.ProjectTwo.utilities.isValidWeight
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.time.LocalDate
import java.time.ZoneId


//@Author Christian Clark
//@Date 1-9-26

/* This is the fragment that contains the list of recoded weights
** Each record allows for deletion or changing the date
** When a user gets close to, or reaches, their goal, a notification will be sent
*/

class DatabaseGridFragment: Fragment(){

    private lateinit var _adapter: WeightCardAdapter
    private lateinit var _weightList: MutableList<UserInfoEntity>
    private val _userId: Long by lazy { (activity as CoreApp).getId() }

    // https://developer.android.com/topic/libraries/view-binding
    // Use view binding instead of getting by id
    private var _binding: DatabaseGridRecycleFragmentBinding? = null
    private val binding get() = _binding!!
    // (Google, 2025)
    private var notificationIdCurrent: Int = 0
    // Google. (2025, February 10). View binding.
    // Retrieved from Android Developers: https://developer.android.com/topic/libraries/view-binding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceData: Bundle?): View {
        _binding = DatabaseGridRecycleFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        lifecycleScope.launch {
            initData()

            setupRecyclerView()
            setupAddButton()
        }
    }

    suspend private fun initData(){
        try{
            withContext(Dispatchers.IO){
                val db = WeightDatabase.getInstance(requireContext()).userInfoDao()
                _weightList = db.GetInfoListByUser(_userId).toMutableList()
            }
        }catch(_: Exception){
            Toast.makeText(context, getString(R.string.could_not_get_user_data), Toast.LENGTH_LONG).show()
        }
    }

    suspend private fun setupAddButton(){
        binding.weightAddButton.setOnClickListener { _ ->
            lifecycleScope.launch {
                try{
                    var currWeight = binding.weightEntryDatabase.text.toString().toFloatOrNull()
                    binding.weightEntryDatabase.setText("")
                    if(currWeight == null || !currWeight.isValidWeight()){
                        Toast.makeText(context, getString(R.string.error_invalid_weight), Toast.LENGTH_SHORT).show()
                        return@launch
                    }

                    val currDate = System.currentTimeMillis()
                    try{
                        withContext(Dispatchers.IO){
                            val db = WeightDatabase.getInstance(requireContext())
                            AddNewWeight(_userId, currDate, currWeight, db.userInfoDao())
                            _weightList = db.userInfoDao().GetInfoListByUser(_userId).toMutableList()
                        }
                        _adapter.updateData(_weightList)
                    }catch (_: Exception){
                        Toast.makeText(context,
                            getString(R.string.could_not_add_new_weight), Toast.LENGTH_SHORT).show()
                        return@launch
                    }

                    var goalWeightRow: GoalEntity?
                    var goalWeight: Float? = null
                    try{
                        goalWeightRow = withContext(Dispatchers.IO){
                            WeightDatabase.getInstance(requireContext()).goalDao().GetGoalRow(_userId)
                        }
                        if(goalWeightRow == null){
                            throw Exception()
                        }
                        goalWeight = goalWeightRow.goal
                    }catch (_: Exception){
                        Toast.makeText(context,
                            getString(R.string.goal_weight_could_not_be_accessed), Toast.LENGTH_SHORT).show()
                    }
                    if (goalWeight == null){
                        Toast.makeText(context, getString(R.string.goal_weight_could_not_be_accessed),
                            Toast.LENGTH_SHORT).show()
                        return@launch
                    }
                    if(_weightList.isEmpty()){
                        return@launch
                    }
                    currWeight = _weightList.get(0).weight
                    var weightToGo = 0f
                    var alert: AlertType = AlertType.DO_NOT_SEND
                    if(currWeight > goalWeight){
                        weightToGo = currWeight - goalWeight
                        if(weightToGo <= 10.0f){
                            alert = AlertType.APPROACHING
                        }
                    }else if(currWeight <= goalWeight){
                        alert = AlertType.REACHED_GOAL
                    }

                    if(alert != AlertType.DO_NOT_SEND){
                        attemptSendNotification(alert, weightToGo)
                    }

                }catch (_: Exception){
                    Toast.makeText(requireContext(), getString(R.string.error_invalid_weight), Toast.LENGTH_SHORT).show()
                }

            }

        }
    }

    private fun setupRecyclerView(){
        binding.recyclerView.layoutManager = LinearLayoutManager(context)

        _adapter = WeightCardAdapter(requireContext(), _weightList, _userId,
            ::deleteClicked,
            ::dateClicked)

        binding.recyclerView.adapter = _adapter
        binding.recyclerView.setHasFixedSize(true)
    }

    // removes an entry from the database when the delete button is tapped
    fun deleteClicked(info: UserInfoEntity, position: Int){
        lifecycleScope.launch {
           withContext(Dispatchers.IO){
               val db = WeightDatabase.getInstance(requireContext()).userInfoDao()
               db.RemoveWeightAt(info.uid)
           }
            _adapter.RemoveItem(position)
        }
    }

    fun dateClicked(info: UserInfoEntity, position: Int){
        val date = LocalDate.now()
        DatePickerDialog(
            requireContext(), { _, year, month, dayOfMonth ->
                lifecycleScope.launch {
                    val date: Long = LocalDate.of(year, month + 1, dayOfMonth).atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli()
                    val list = withContext(Dispatchers.IO){
                        val db = WeightDatabase.getInstance(requireContext()).userInfoDao()
                        val copy = info.copy(date = date)
                        db.ChangeDateAt(copy)
                        db.GetInfoListByUser(_userId)
                    }
                    _adapter.updateData(list)
                }

            }, date.year, date.monthValue - 1, date.dayOfMonth
        ).show()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    // chooses which notification to send
    fun attemptSendNotification(type: AlertType, poundsRemaining: Float){
        if(ContextCompat.checkSelfPermission(requireContext(), Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED){
            when (type){
                AlertType.APPROACHING -> sendNotification(getString(R.string.approaching_title), getString(R.string.approaching_message, poundsRemaining))
                AlertType.REACHED_GOAL -> sendNotification(getString(R.string.reached_title), getString(R.string.reached_message))
                else -> return
            }
        }
    }

    // sends a notification and updates the notification id
    fun sendNotification(title: String, message: String){
        val notiBuilder = NotificationCompat.Builder(requireContext(), "goal_alerts")
            .setSmallIcon(R.drawable.arrow)
            .setContentTitle(title)
            .setContentText(message)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)

        val notiManager = requireContext().getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notiManager.notify(notificationIdCurrent, notiBuilder.build())
        notificationIdCurrent++
    }

    // enum class to determine which notification to send
    enum class AlertType{
        APPROACHING,
        REACHED_GOAL,
        DO_NOT_SEND
    }
}
